fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -1940,6 +1844,172 @@ struct RocksDBKeyValueStore : IKeyValueStore {
 	}
 };
 
+void RocksDBKeyValueStore::Writer::action(CheckpointAction& a) {
+	TraceEvent("RocksDBServeCheckpointBegin", id)
+	    .detail("MinVersion", a.request.version)
+	    .detail("Range", a.request.range.toString())
+	    .detail("Format", static_cast<int>(a.request.format))
+	    .detail("CheckpointDir", a.request.checkpointDir);
+
+	rocksdb::Checkpoint* checkpoint;
+	rocksdb::Status s = rocksdb::Checkpoint::Create(db, &checkpoint);
+	if (!s.ok()) {
+		logRocksDBError(s, "Checkpoint");
+		a.reply.sendError(statusToError(s));
+		return;
+	}
+
+	rocksdb::PinnableSlice value;
+	rocksdb::ReadOptions readOptions = getReadOptions();
+	s = db->Get(readOptions, cf, toSlice(persistVersion), &value);
+
+	if (!s.ok() && !s.IsNotFound()) {
+		logRocksDBError(s, "Checkpoint");
+		a.reply.sendError(statusToError(s));
+		return;
+	}
+
+	const Version version =
+	    s.IsNotFound() ? latestVersion : BinaryReader::fromStringRef<Version>(toStringRef(value), Unversioned());
+
+	ASSERT(a.request.version == version || a.request.version == latestVersion);
+	TraceEvent(SevDebug, "RocksDBServeCheckpointVersion", id)
+	    .detail("CheckpointVersion", a.request.version)
+	    .detail("PersistVersion", version);
+
+	// TODO: set the range as the actual shard range.
+	CheckpointMetaData res(version, a.request.range, a.request.format, a.request.checkpointID);
+	const std::string& checkpointDir = a.request.checkpointDir;
+
+	if (a.request.format == RocksDBColumnFamily) {
+		rocksdb::ExportImportFilesMetaData* pMetadata;
+		platform::eraseDirectoryRecursive(checkpointDir);
```

Comment:
Where does checkpointDir come from?  (Can a client make it something like "../../" or some such thing?)

Reply:
The checkpoint dir is decided in ACTOR `createCheckpoint()` at `storageserver.actor.cpp:6988`.

Added `abspath()` here to resolve potential `../..`, but currently the path is decided only by a storage server.